modules/gapi/src/compiler/gcompiler.cpp

```diff
@@ -317,14 +318,31 @@ void cv::gimpl::GCompiler::validateInputMeta()
                      "got " + std::to_string(m_metas.size()) + " meta arguments)"));
     }
 
-    const auto meta_matches = [](const GMetaArg &meta, const GProtoArg &proto) {
+    const auto meta_matches = [](const GMetaArg &meta, const GProtoArg &proto, std::ostream& tracer) {
         switch (proto.index())
         {
         // FIXME: Auto-generate methods like this from traits:
         case GProtoArg::index_of<cv::GMat>():
         case GProtoArg::index_of<cv::GMatP>():
-            return util::holds_alternative<cv::GMatDesc>(meta);
+        {
+            // check actual type
+            if (!util::holds_alternative<cv::GMatDesc>(meta))
+            {
+                return false;
+            }
 
+            // check meta-data on validity
+            try
+            {
+                cv::validate_input_meta(cv::util::get<cv::GMatDesc>(meta)); //may throw
+            }
+            catch (const std::exception& ex)
+            {
+                tracer << ex.what();
+                return false;
+            }
```

Comment:
Hmm so I get your idea now and y u didn't want to use exceptions - ok

This `meta_matches` function previously checked only for a _match_ - if the right type is stored in variant for the given graph data type, and the below text was complaining only on a mismatch. In this case, returning `bool` was ok as it was just a switch over "is x of type X" collection of checks. 

Now we also look inside and it is not only about matching but also about data be valid... and we check if the data is valid somewhere inside.

Soo... I'd propose to separate these two checks to lay out the code more correctly. Please keep the function in its old form (making only "is" checks):

```c++
    const auto meta_matches = [](const GMetaArg &meta, const GProtoArg &proto) {
        switch (proto.index())
        {
        // FIXME: Auto-generate methods like this from traits:
        case GProtoArg::index_of<cv::GMat>():
        case GProtoArg::index_of<cv::GMatP>():
            return util::holds_alternative<cv::GMatDesc>(meta);

        case GProtoArg::index_of<cv::GFrame>():
            return util::holds_alternative<cv::GFrameDesc>(meta);

        case GProtoArg::index_of<cv::GScalar>():
            return util::holds_alternative<cv::GScalarDesc>(meta);

        case GProtoArg::index_of<cv::detail::GArrayU>():
            return util::holds_alternative<cv::GArrayDesc>(meta);

        case GProtoArg::index_of<cv::detail::GOpaqueU>():
            return util::holds_alternative<cv::GOpaqueDesc>(meta);

        default:
            GAPI_Assert(false);
        }
        return false; // should never happen
    };
```
so then you don't need to change the below exception code on meta MISMATCH.

After you checked for meta MISMATCH, call your new routine (which throws - just like that) to check if the meta is itself correct.

Reply:
@dmatveev 
Yes I thought about separation already and this is a reason why i have started visitor implementation to do not duplicate switch-cases
ok, add new lambda here